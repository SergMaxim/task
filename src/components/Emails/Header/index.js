import React from 'react';
import Styles from './header.scss';

const Header = () => (
    <section className = { Styles.header }>
        <ul>
            <li>
                <a>Home</a>
            </li>
            <li>Email</li>
        </ul>
        <p>Email</p>
    </section>
);

export default Header;
