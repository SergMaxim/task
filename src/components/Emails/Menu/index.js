import React, { Component } from 'react';
import Styles from './styles.scss';
import { func, object } from 'prop-types';

export default class Menu extends Component {

    /*static propTypes = {
        menu:     object.isRequired,
        openForm: func.isRequired
    };*/
    render () {
        return (
            <section className = { Styles.menu }>
                <div className = { Styles.button }>Compose Mail</div>
            </section>
        );
    }
}
