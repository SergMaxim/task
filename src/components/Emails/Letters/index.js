import React, { Component } from 'react';
import { arrayOf, bool, shape, string, number, func } from 'prop-types';
import Styles from './styles.scss';
import Letter from './Letter';
import Header from './Header';
import { getUniqueID } from '../../../helpers';

export default class Letters extends Component {
    static propTypes = {
        letters: arrayOf(
            shape({
                name:  string.isRequired,
                star:  bool.isRequired,
                theme: string.isRequired,
                time:  number.isRequired
            }).isRequired
        ).isRequired,
        searchLetters: func.isRequired
    };

    render () {
        const { letters: postsData, searchLetters } = this.props;
        const letters = postsData.map((props) => {
            const id = getUniqueID();

            return <Letter key = { id } { ...props } />;
        });

        return (
            <section className = { Styles.letters }>
                <Header
                    count = { postsData.length }
                    searchLetters = { searchLetters }
                />
                {letters}
            </section>
        );
    }
}
