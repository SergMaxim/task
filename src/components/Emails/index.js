// Core
import React, { Component } from 'react';
import { getLetters } from '../../helpers/fetch';

import Header from './Header';
import Letters from './Letters';
import Menu from './Menu';
import Styles from './styles.scss';

export default class Emails extends Component {
    constructor () {
        super();
        this.openForm = ::this._openForm;
        this.getLetters = ::this._getLetters;
        this.searchLetters = ::this._searchLetters;
    }

    state = {
        isForm:        false,
        currentLetter: 0,
        letters:       [],
        lettersBuf:    []
    };

    componentWillMount () {
        this.getLetters();
    }
    _openForm () {
        this.setState({
            isForm: !this.state.isForm
        });
    }

    _searchLetters (comment) {
        const letters = this.state.lettersBuf.filter(
            (value) =>
                value.name.indexOf(comment) >= 0 ||
                value.theme.indexOf(comment) >= 0
        );

        this.setState(() => ({
            letters
        }));
    }

    async _getLetters () {
        const { letters } = await getLetters('/letters.json');

        this.setState(() => ({
            letters,
            lettersBuf: letters
        }));
    }
    render () {
        const { letters } = this.state;

        return (
            <section className = { Styles.emails }>
                <Header />
                <div className = { Styles.emailsBody }>
                    <Menu />
                    <Letters
                        letters = { letters }
                        searchLetters = { this.searchLetters }
                    />
                </div>
            </section>
        );
    }
}
