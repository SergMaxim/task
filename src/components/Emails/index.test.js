import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import dom from 'react-test-renderer';

// Instruments
import Emails from './index';

configure({ adapter: new Adapter() });

global.fetch = jest.fn(() => ({
    data: [
        {
            name:  'Mansion Tech',
            theme: 'New Tasks',
            body:  'New Tasks',
            star:  true,
            type:  1,
            time:  1513545795
        },
        {
            name:  'Mansion Tech',
            theme: 'New Tasks',
            body:  'New Tasks',
            star:  true,
            type:  1,
            time:  1513545795
        }
    ],
    message: 'the request has succeeded'
}));

const classEmails = new Emails;
const result = mount(<Emails />);
const renderTree = dom.create(<Emails />).toJSON();
const message = 'message';

describe('Emails component', () => {
    test('Emails component should correspond to its snapshot counterpart', () => {
        expect(renderTree).toMatchSnapshot();
    });

    test('searchLetters function shoud be s function', () => {
        expect(typeof classEmails.searchLetters).toBe('function');
    });

    test(`Letters head  state and input value should reflect according changes if any text input provided`, () => {
        result.find('input[type=\'text\']').simulate('keyUp', {
            target: {
                value: message
            }
        });

        expect(result.state().letters).toEqual([]);
    });

});
