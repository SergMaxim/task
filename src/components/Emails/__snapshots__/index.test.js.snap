// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emails component Emails component should correspond to its snapshot counterpart 1`] = `
<section
  className="emails"
>
  <section
    className="header"
  >
    <ul>
      <li>
        <a>
          Home
        </a>
      </li>
      <li>
        Email
      </li>
    </ul>
    <p>
      Email
    </p>
  </section>
  <div
    className="emailsBody"
  >
    <section
      className="menu"
    >
      <div
        className="button"
      >
        Compose Mail
      </div>
    </section>
    <section
      className="letters"
    >
      <section
        className="header"
      >
        <div>
          Inbox 
          <span>
            0
          </span>
        </div>
        <form
          className="search"
          onSubmit={[Function]}
        >
          <input
            onKeyUp={[Function]}
            placeholder="search"
            type="text"
          />
          <input
            type="submit"
            value=""
          />
        </form>
      </section>
    </section>
  </div>
</section>
`;
